### PR TITLE
W3t close channel

### PR DESCRIPTION
**Pull request by spersico**
This PR:
- Prevents the web3torrentClient to trigger payments while a torrent is being cancelled (by pausing the torrent):
https://github.com/statechannels/monorepo/compare/w3t-close-channel?expand=1#diff-a84153ed9f3ab2bd13c41bca43b584c7R134
https://github.com/statechannels/monorepo/compare/w3t-close-channel?expand=1#diff-a84153ed9f3ab2bd13c41bca43b584c7L401
- moves the logic of waiting for the user's turn before closing to the PaymentChannelClient:
https://github.com/statechannels/monorepo/compare/w3t-close-channel?expand=1#diff-9b0f77e1a02c15a73ad86b3e5d0b6620R212
- Makes channelClose calls independent of each other:
https://github.com/statechannels/monorepo/compare/w3t-close-channel?expand=1#diff-a84153ed9f3ab2bd13c41bca43b584c7R522-R525

related to #1961 and #1918 and #1811

### CODE REVIEW

**Comment by snario** [comment]
I found this: https://30317-209829093-gh.circle-artifacts.com/0/tmp/seed-download-cancel.pino.log

```json
{
   "time":1590450980252,
   "level":30,
   "jsonRpcResponse":{
      "id":12,
      "jsonrpc":"2.0",
      "error":{
         "code":300,
         "message":"Not your turn"
      }
   },
   "msg":"OUTGOING JSONRPC REQUEST:",
   "name":"xstate-wallet",
   "browserId":1
}
```

**Comment by spersico** [comment]
> I found [this log](https://30317-209829093-gh.circle-artifacts.com/0/tmp/seed-download-cancel.pino.log) from the [failed `integration-test-web3torrent-optional` job](https://app.circleci.com/pipelines/github/statechannels/monorepo/6862/workflows/428be103-e527-44f4-985b-7ce8144bdeb8/jobs/30317/artifacts)
> 
> ```json
> {
>    "time":1590450980252,
>    "level":30,
>    "jsonRpcResponse":{
>       "id":12,
>       "jsonrpc":"2.0",
>       "error":{
>          "code":300,
>          "message":"Not your turn"
>       }
>    },
>    "msg":"OUTGOING JSONRPC REQUEST:",
>    "name":"xstate-wallet",
>    "browserId":1
> }
> ```

yeah, I paired with @andrewgordstewart last night and made with him this PR, that puts the ball in xstate-wallet in regards with this error, so that he can start hacking into it on that side.
What I mean with this is: this PR is not supposed to fix the error, but it solves the issues related to it in web3torrent's side

**Comment by snario** [comment]
@spersico Does this PR _fix_ #1961?

File: packages/web3torrent/src/clients/payment-channel-client.ts [hunk 1]
```diff
@@ -209,6 +209,8 @@ export class PaymentChannelClient {
       channelState.channelId === channelId && channelState.status === 'closed';
 
     if (!['closing', 'closed'].includes(this.channelCache[channelId].status)) {
+      await this.getLatestPaymentReceipt(channelId);
+      log.info('My turn, about to close channel');
```
Comment (andrewgordstewart) [inline packages/web3torrent/src/clients/payment-channel-client.ts:213]: This doesn't look like it's intended to get into master.

File: packages/web3torrent/src/library/web3torrent-lib.ts [hunk 2]
```diff
@@ -130,6 +131,7 @@ export default class WebTorrentPaidStreamingClient extends WebTorrent {
     log.info('> Cancelling download. Closing payment channels, and then removing torrent');
     const torrent = this.torrents.find(t => t.infoHash === infoHash);
     if (torrent) {
+      torrent.pause();
```
Comment (andrewgordstewart) [inline packages/web3torrent/src/library/web3torrent-lib.ts:134]: Maybe add a bit of explanation, since the intended effect is a bit subtle?
```suggestion
      // I am only allowed to close the channel on my turn.
      // If I don't pause the torrent, then I will continue to make payments, meaning the call to 
      // this.closeChannels would race the `makePayment` function, and one of them would fail with a
      // "Not my turn" error.
      torrent.pause();
```

File: packages/web3torrent/src/library/web3torrent-lib.ts [hunk 3]
```diff
@@ -518,19 +518,19 @@ export default class WebTorrentPaidStreamingClient extends WebTorrent {
...
-          log.info(`About to close payment channel ${leechingChannelId}`);
-          await this.paymentChannelClient.getLatestPaymentReceipt(leechingChannelId);
-          await this.paymentChannelClient.closeChannel(leechingChannelId);
-          wire.paidStreamingExtension.leechingChannelId = null;
-          log.info(`Channel ${leechingChannelId} closed`);
+          log.info(`SP-> About to close payment channel ${leechingChannelId}`);
+          return this.paymentChannelClient.closeChannel(leechingChannelId, true).then(() => {
+            wire.paidStreamingExtension.leechingChannelId = null;
+            log.info(`SP-> Channel closed: ${leechingChannelId}`);
+          });
```
Comment (andrewgordstewart) [inline packages/web3torrent/src/library/web3torrent-lib.ts]: Why the change from `await` to `.then`?

File: packages/web3torrent/src/library/web3torrent-lib.ts [hunk 4]
```diff
@@ -519,18 +523,18 @@ export default class WebTorrentPaidStreamingClient extends WebTorrent {
         const {seedingChannelId, leechingChannelId} = wire.paidStreamingExtension;
         if (leechingChannelId) {
           log.info(`About to close payment channel ${leechingChannelId}`);
-          await this.paymentChannelClient.getLatestPaymentReceipt(leechingChannelId);
-          await this.paymentChannelClient.closeChannel(leechingChannelId);
-          wire.paidStreamingExtension.leechingChannelId = null;
-          log.info(`Channel ${leechingChannelId} closed`);
+          return this.paymentChannelClient.closeChannel(leechingChannelId).then(() => {
```
Comment (kerzhner) [inline packages/web3torrent/src/library/web3torrent-lib.ts:526]: I think @andrewgordstewart already asked this, but is there a reason to use `.then` promise syntax here instead of await?

File: packages/web3torrent/src/library/web3torrent-lib.ts [hunk 5]
```diff
@@ -519,18 +523,18 @@ export default class WebTorrentPaidStreamingClient extends WebTorrent {
...
+            wire.paidStreamingExtension.leechingChannelId = null;
+            log.info(`Channel closed: ${leechingChannelId}`);
+          });
```
Comment (snario) [inline packages/web3torrent/src/library/web3torrent-lib.ts:529]: Why the switch to `.then`?

File: packages/web3torrent/src/library/web3torrent-lib.ts [hunk 6]
```diff
@@ -130,6 +131,11 @@ export default class WebTorrentPaidStreamingClient extends WebTorrent {
     log.info('> Cancelling download. Closing payment channels, and then removing torrent');
     const torrent = this.torrents.find(t => t.infoHash === infoHash);
     if (torrent) {
+      // I am only allowed to close the channel on my turn.
+      // If I don't pause the torrent, then I will continue to make payments, meaning the call to 
+      // this.closeChannels would race the `makePayment` function, and one of them would fail with a
+      // "Not my turn" error.
+      torrent.pause();
```
Comment (kerzhner) [inline packages/web3torrent/src/library/web3torrent-lib.ts:138]: The [documentation for pausing](https://webtorrent.io/docs) says: Temporarily stop connecting to new peers. Note that this does not pause new incoming connections, nor does it pause the streams of existing connections or their wires.

Given this, is it true that the leecher would stop making payments?

File: packages/web3torrent/src/library/web3torrent-lib.ts [hunk 7]
```diff
@@ -120,6 +120,7 @@ export default class WebTorrentPaidStreamingClient extends WebTorrent {
     const torrent = this.torrents.find(t => t.infoHash === infoHash);
     if (torrent) {
       torrent.pause(); // the paymentChannelClosing is done at the moment of payment
+      this.closeChannels(torrent, true);
```
Comment (kerzhner) [inline packages/web3torrent/src/library/web3torrent-lib.ts:123]: Just to make sure I understand the current application correctly. This `pause` method that includes this line is never called, right?

File: packages/web3torrent/src/library/web3torrent-lib.ts (hunk 4 above)
Comment (spersico) [reply packages/web3torrent/src/library/web3torrent-lib.ts:526]: I wanted to chain promises, but I'll make an extra function and use that instead, so that I maintain consistency.

File: packages/web3torrent/src/library/web3torrent-lib.ts (hunk 5 above)
Comment (spersico) [reply packages/web3torrent/src/library/web3torrent-lib.ts:529]: I wanted to chain promises, but I'll make an extra function and use that instead, so that I maintain consistency.

File: packages/web3torrent/src/library/web3torrent-lib.ts (hunk 3 above)
Comment (spersico) [reply packages/web3torrent/src/library/web3torrent-lib.ts]: I wanted to chain promises, but I'll make an extra function and use that instead, so that I maintain consistency.

File: packages/web3torrent/src/library/web3torrent-lib.ts (hunk 6 above)
Comment (spersico) [reply packages/web3torrent/src/library/web3torrent-lib.ts:138]: Yes, according to the docs it only makes it stop making connections, but in Web3Torrent it also sets the `torrent.pause` to `true` which, in here: https://github.com/statechannels/monorepo/compare/w3t-close-channel?expand=1#diff-a84153ed9f3ab2bd13c41bca43b584c7R404-R405
it makes the leecher stop making payments.
This has the effect that the leecher doesn't pollutes the payment channel with messages when the closing flow is running, and avoids *some* errors on that flow.

File: packages/web3torrent/src/library/web3torrent-lib.ts (hunk 7 above)
Comment (spersico) [reply packages/web3torrent/src/library/web3torrent-lib.ts:123]: The `web3torrent.pause()` method is not being used currently, no.

File: packages/web3torrent/src/library/web3torrent-lib.ts (hunk 4 above)
Comment (kerzhner) [reply packages/web3torrent/src/library/web3torrent-lib.ts:526]: Do we need a separate function? Would this work?
```suggestion
          await this.paymentChannelClient.closeChannel(leechingChannelId);
          wire.paidStreamingExtension.leechingChannelId = null;
          log.info(`Channel closed: ${leechingChannelId}`);```

File: packages/web3torrent/src/library/web3torrent-lib.ts (hunk 5 above)
Comment (andrewgordstewart) [reply packages/web3torrent/src/library/web3torrent-lib.ts:529]: I'm curious why you wanted to chain promises?

Comment (spersico) [reply packages/web3torrent/src/library/web3torrent-lib.ts]: https://github.com/statechannels/monorepo/pull/1969/commits/034e40f790457b167989f3792ebeda37eaecdbb4

File: packages/web3torrent/src/library/web3torrent-lib.ts (hunk 4 above)
Comment (spersico) [reply packages/web3torrent/src/library/web3torrent-lib.ts]: https://github.com/statechannels/monorepo/pull/1969/commits/034e40f790457b167989f3792ebeda37eaecdbb4

File: packages/web3torrent/src/library/web3torrent-lib.ts (hunk 3 above)
Comment (spersico) [reply packages/web3torrent/src/library/web3torrent-lib.ts]: https://github.com/statechannels/monorepo/pull/1969/commits/034e40f790457b167989f3792ebeda37eaecdbb4

File: packages/web3torrent/src/clients/payment-channel-client.ts (hunk 1 above)
Comment (spersico) [reply packages/web3torrent/src/clients/payment-channel-client.ts]: https://github.com/statechannels/monorepo/pull/1969/commits/034e40f790457b167989f3792ebeda37eaecdbb4

File: packages/web3torrent/src/library/web3torrent-lib.ts (hunk 5 above)
Comment (spersico) [reply packages/web3torrent/src/library/web3torrent-lib.ts]: > I'm curious why you wanted to chain promises?

Because the seeding or leeching `channelId` should only be closed AFTER the channel has been closed (because it's used by the communication flow). By chaining promises, we can launch `channelClose` for all wires at the same time (with an array and using the Promise.all), which results in a faster torrent cancelling.

File: packages/web3torrent/src/library/web3torrent-lib.ts (hunk 4 above)
Comment (spersico) [reply packages/web3torrent/src/library/web3torrent-lib.ts]: > Do we need a separate function? Would this work?

this would close the channel one by one, unless i'm wrong.

File: packages/web3torrent/src/library/web3torrent-lib.ts (hunk 5 above)
Comment (andrewgordstewart) [reply packages/web3torrent/src/library/web3torrent-lib.ts]: Hmm, I'm not sure I follow. Two awaits in a row is semantically equivalent to one promise plus one `then`, isn't it?

Comment (spersico) [reply packages/web3torrent/src/library/web3torrent-lib.ts]: IIRC no:
```
await Promise()
await Promise()
```
is not equivalent to
```
await Promise().then(()=>Promise())
```
but to
```
async ()=>{
  await Promise()
  await Promise()
}
```

And for a `Promise.all` I understand that you need an array of single promises...
..
...
now that I think about it, in this case, you might be right... well, it's been changed already.
